### PR TITLE
Added Integration tests

### PR DESCRIPTION
**Pull request by addb**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/xdm/XDMMediaEventType.kt [hunk 1]
```diff
@@ -35,5 +35,6 @@ internal enum class XDMMediaEventType(val value: String) {
         fun getTypeString(type: XDMMediaEventType): String {
             return "media.${type.value}"
         }
+        fun getByValue(value: String) = XDMMediaEventType.values().find { it.value == value }
```
Comment (kevinlind) [inline code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/xdm/XDMMediaEventType.kt:38]: I don't see this function being used?  Is it still needed?

**Comment by kevinlind** [comment]
@addb It looks like you copied several test util files from other repos, such as under the `util` and `services` folders. Can you add comments to the files which you copied and mention what changes you made to them, if any? It will help me know where to put my focus and which files are essentially boilerplate for the functional test framework.

**Comment by kevinlind** [comment]
@addb Can you update the `.circleci/config.yml` CircleCI configuration to store the functional test results?  I believe the command should be:
```
- store_test_results:
          path:  code/edgemedia/build/outputs/androidTest-results
```

File: code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/edge/media/internal/MediaEdgeLocationHintIntegrationTests.kt [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            latch.countDown()
+        }
+
+        latch.await()
+        FunctionalTestHelper.assertExpectedEvents(false)
+        FunctionalTestHelper.resetTestExpectations()
+    }
+
+    @Test
+    fun testMediaEdgeRequests_whenLocationHintSet_urlPathContainsLocationHint() {
```
[the remark this answers is not in the document]
Comment (addb) [reply code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/edge/media/internal/MediaEdgeLocationHintIntegrationTests.kt:84]: Yes the other file has no location hint but I dont verify the URL, so I could just add one more test here.

File: code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/edge/media/internal/MediaEdgeIntegrationTests.kt [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+
+        assertXDMData(networkRequests[0], XDMMediaEventType.SESSION_START, mediaInfo, metadata, configuration)
+        assertXDMData(networkRequests[1], XDMMediaEventType.PLAY, backendSessionId = testBackendSessionId)
+        assertXDMData(networkRequests[2], XDMMediaEventType.PAUSE_START, backendSessionId = testBackendSessionId, playhead = 7)
+        assertXDMData(networkRequests[3], XDMMediaEventType.SESSION_COMPLETE, backendSessionId = testBackendSessionId, playhead = 7)
+    }
+
+    @Test
+    fun testSessionStartErrorResponse_shouldNotSendAnyOtherNetworkRequests() {
+        val responseConnection = createNetworkResponse(ERROR_RESPONSE_STRING, 200)
```
[the remark this answers is not in the document]
Comment (addb) [reply code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/edge/media/internal/MediaEdgeIntegrationTests.kt:135]: Let me check that.

**Comment by addb** [comment]
> @addb It looks like you copied several test util files from other repos, such as under the `util` and `services` folders. Can you add comments to the files which you copied and mention what changes you made to them, if any? It will help me know where to put my focus and which files are essentially boilerplate for the functional test framework.

Yes, so I have just added logic to store all networkRequests and a getter for our use case similar to iOS. So 

These 2 are the new functions:
FunctionalTestHelper.getAllNetworkRequests
FunctionalTestNetworkService.getAllReceivedTestableNetworkRequests

File: code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/util/FunctionalTestHelper.java [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+    }
+
+    /**
+     * Use this API for JSON formatted {@code NetworkRequest} body in order to retrieve a flattened
+     * map containing its data.
+     *
+     * @param networkRequest the {@link NetworkRequest} to parse
+     * @return The JSON request body represented as a flatten map
+     */
+    public static Map<String, String> getFlattenedNetworkRequestBody(
```
Comment (kevinlind) [inline code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/util/FunctionalTestHelper.java:884]: It doesn't look like you are currently using this function to get a "flattened" network request body. In your tests, you are getting the request body as a Map<String, Any> and doing the comparison that way, which works better in this case. However, this method is the only place where I see the FunctionalTestUtils being used. Unless you think we will need this function in the future, I'd like to remove it along with the entire FunctionalTestUtils class to remove the dead code.

File: code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/edge/media/internal/MediaEdgeIntegrationTests.kt [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+@RunWith(AndroidJUnit4::class)
+class MediaEdgeIntegrationTests {
+    companion object {
+        private const val EDGE_CONFIG_ID = "1234abcd-abcd-1234-5678-123456abcdef"
+        private const val SESSION_START_URL = "https://edge.adobedc.net/ee/va/v1/sessionStart"
+        private const val testBackendSessionId = "99cf4e3e7145d8e2b8f4f1e9e1a08cd52518a74091c0b0c611ca97b259e03a4d"
+        private const val SUCCESS_RESPONSE_STRING =
+            "\u0000{\"handle\":[{\"payload\":[{\"sessionId\":\"99cf4e3e7145d8e2b8f4f1e9e1a08cd52518a74091c0b0c611ca97b259e03a4d\"}],\"type\":\"media-analytics:new-session\",\"eventIndex\":0}]}"
+        private const val ERROR_RESPONSE_STRING =
+            "\u0000{\"errors\" : [{\"type\" : \"https://ns.adobe.com/aep/errors/va-edge-0404-404\", \"status\" : 404,\"title\" : \"Not Found\", \"detail\" : \"The requested resource could not be found but may be available again in the future.\",\"report\" : {\"details\" : \"Error processing request. If the session is longer than 24h, please start a new one. Returning Not Found\"}}]}"
```
[the remark this answers is not in the document]
Comment (kevinlind) [reply code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/edge/media/internal/MediaEdgeIntegrationTests.kt:55]: Can you make the change to use `va-edge-0400-400` and `400`?

File: code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/util/FunctionalTestUtils.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import org.json.JSONObject;
+
+public class FunctionalTestUtils {
+
+    private static final String LOG_SOURCE = "FunctionalTestUtils";
+
+    private FunctionalTestUtils() {}
```
Comment (kevinlind) [inline code/edgemedia/src/androidTest/java/com/adobe/marketing/mobile/util/FunctionalTestUtils.java:33]: Can this class be removed now? I don't see that it is being used.

**Review by kevinlind — APPROVED** [review]

**Pull request merged**